File: web/src/api/system/menu.ts
import { request } from "@/utils/service"

export interface MenusData {
  id: number
  pid: number
  name: string
  path: string
  redirect?: string
  component: string
  meta: {
    hidden?: boolean
    title: string
    elIcon?: string
    svgIcon?: string
    affix?: boolean
    keepAlive?: boolean
  }
  children?: MenusData[]
}

type MenusResponseData = ApiResponseData<MenusData[]>

// 获取动态路由
export function getMenus() {
  return request<MenusResponseData>({
    url: "/menu/getMenus",
    method: "get"
  })
}

interface reqMenu {
  pid: number
  name?: string
  path: string
  redirect?: string
  component: string
  meta: {
    hidden?: boolean
    title?: string
    icon?: string
    affix?: boolean
    keepAlive?: boolean
  }
}

export function addMenuApi(data: reqMenu) {
  return request<ApiResponseData<null>>({
    url: "menu/addMenu",
    method: "post",
    data
  })
}

interface editReq extends reqMenu {
  id: number
}

export function editMenuApi(data: editReq) {
  return request<ApiResponseData<null>>({
    url: "menu/editMenu",
    method: "post",
    data
  })
}

export function deleteMenuApi(data: reqId) {
  return request<ApiResponseData<null>>({
    url: "menu/deleteMenu",
    method: "post",
    data
  })
}

interface allMenus {
  list: MenusData[]
  menuIds: number[]
}

export function getElTreeMenusApi(data: reqId) {
  return request<ApiResponseData<allMenus>>({
    url: "menu/getElTreeMenus",
    method: "post",
    data
  })
}

File: server/model/system/request/menu.go
package request

type Menu struct {
	Pid       uint   `json:"pid"` // 默认0 根目录
	Name      string `json:"name"`
	Path      string `json:"path" validate:"required"`
	Redirect  string `json:"redirect"`
	Component string `json:"component" validate:"required"`
	Sort      uint   `json:"sort" validate:"required"`
	Meta      meta   `json:"meta"`
}

type meta struct {
	Hidden    bool   `json:"hidden"` // 菜单是否隐藏
	Title     string `json:"title"`  // 菜单名
	Icon      string `json:"icon"`   // element图标
	Affix     bool   `json:"affix"`  // 是否固定
	KeepAlive bool   `json:"keepAlive"`
}

type EditMenuReq struct {
	Id uint `json:"id" validate:"required"`
	Menu
}

type EditRoleMenu struct {
	RoleId uint   `json:"roleId"`
	Ids    []uint `json:"ids"`
}
